src/ws/session.c

```diff
@@ -550,17 +551,18 @@ perform_tlscert (const char *rhost)
 
   debug ("start tls-cert authentication for cockpit-ws %u", getppid ());
 
-  /* pam_cockpit_cert sets the user name from the certificate */
-  res = pam_start ("cockpit", NULL, &conv, &pamh);
+  char *username = cockpit_session_client_certificate_check ();
+  if (username == NULL)
+    exit_init_problem (PAM_AUTH_ERR);
+
+  res = pam_start ("cockpit", username, &conv, &pamh);
```

Comment:
Here would be a good place to free `username` again. I suppose our valgrind check doesn't go across this suid binary, and it's short-lived, so maybe we can just ignore the leak.

Reply:
This was an oversight.